### PR TITLE
Story/csi 1276 svc perf get host (reconsider open issues CSI-1492 as marked in conversations)

### PR DESCRIPTION
**Pull request by igorgonibm**
Improve performance of getting host by identifiers for SVC

### CODE REVIEW

File: controller/array_action/array_mediator_svc.py [hunk 1]
```diff
@@ -294,6 +297,51 @@ def get_host_by_host_identifiers(self, initiators):
...
+            wwns = host_details.get_as_list(HOST_WWPNS_PARAM)
+            host = Host(host_id, host_name, iscsi_names, wwns)
+            res.append(host)
+        return res
+
+    def _get_detailed_hosts_list_cmd(self, host_list):
+        writer = StringIO()
+        for host in host_list:
+            host_id = host.get(HOST_ID_PARAM)
+            writer.write(LIST_HOSTS_CMD)
```
Comment (oriyarde) [inline controller/array_action/array_mediator_svc.py:331]: I think something like
```
LIST_HOSTS_CMD_FORMAT = 'lshost {HOST_ID};'
writer.write(LIST_HOSTS_CMD_FORMAT.format(HOST_ID=host_id)
```
would be more readable and with the same "big O notation" efficiency

File: controller/array_action/array_mediator_svc.py [hunk 2]
```diff
@@ -294,6 +297,51 @@ def get_host_by_host_identifiers(self, initiators):
             logger.debug("can not found host by using initiators: {0} ".format(initiators))
             raise controller_errors.HostNotFoundError(initiators)
 
+    def _get_detailed_hosts_list(self):
+        logger.debug("Getting detailed hosts list on array {0}".format(self.endpoint))
+        hosts_list = self.client.svcinfo.lshost()
+        if not hosts_list:
+            return []
+
+        # for each host get its detailed info from array by sending batch of commands
```
Comment (oriyarde) [inline controller/array_action/array_mediator_svc.py:306]: this might imply that we're sending a batch "for each host".
maybe:
`get all hosts details by sending a single batch of commands, in which each command is per host`?

File: controller/array_action/array_mediator_svc.py [hunk 3]
```diff
@@ -24,6 +27,14 @@
 POOL_NOT_MATCH_VOL_CAPABILITIES = 'CMMVC9292E'
 NOT_REDUCTION_POOL = 'CMMVC9301E'
 
+LIST_HOSTS_CMD = "lshost"
+LIST_CMDS_SEPARATOR = ";"
+HOST_ID_PARAM = 'id'
+HOST_NAME_PARAM = 'name'
+HOST_ISCSI_NAMES_PARAM = 'iscsi_name'
+HOST_WWPNS_PARAM = 'WWPN'
+MAX_HOSTS_LIST_ERR_MSG_LENGTH = 300
```
Comment (oriyarde) [inline controller/array_action/array_mediator_svc.py:36]: `HOSTS_LIST_ERR_MSG_MAX_LENGTH`?

File: controller/array_action/array_mediator_svc.py [hunk 4]
```diff
@@ -294,6 +297,51 @@ def get_host_by_host_identifiers(self, initiators):
...
+    def _get_detailed_hosts_list(self):
+        logger.debug("Getting detailed hosts list on array {0}".format(self.endpoint))
+        hosts_list = self.client.svcinfo.lshost()
+        if not hosts_list:
+            return []
+
+        # for each host get its detailed info from array by sending batch of commands
+        detailed_hosts_list_cmd = self._get_detailed_hosts_list_cmd(hosts_list)
+        logger.debug("Sending getting detailed hosts list commands batch")
+        detailed_hosts_list_output, detailed_hosts_list_errors = self.client.send_raw_command(detailed_hosts_list_cmd)
```
Comment (oriyarde) [inline controller/array_action/array_mediator_svc.py:309]: `send_raw_command` needs a wrapper that would convert the bytes to string for both stdout and stderr

File: controller/array_action/array_mediator_svc.py [hunk 5]
```diff
@@ -294,6 +297,51 @@ def get_host_by_host_identifiers(self, initiators):
...
+            host_id = host.get(HOST_ID_PARAM)
+            writer.write(LIST_HOSTS_CMD)
+            writer.write(" ")
+            writer.write(host_id)
+            writer.write(LIST_CMDS_SEPARATOR)
+        return writer.getvalue()
+
+    def _get_formatted_hosts_list_error_msg(self, detailed_host_list_errors_bytes):
+        detailed_host_list_errors = bytes_to_string(detailed_host_list_errors_bytes)
+        if not detailed_host_list_errors:
```
Comment (oriyarde) [inline controller/array_action/array_mediator_svc.py:339]: no need for this `if`

File: controller/array_action/svc_cli_result_reader.py [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+import controller.array_action.errors as controller_errors
+
+logger = get_stdout_logger()
+
+ID_PARAM_NAME = "id"
+
+
+class SVCListResultsReader:
+    """
+    Iterable object used to read raw command response from m SVS array.
```
Comment (roysahar-ibm) [inline controller/array_action/svc_cli_result_reader.py:11]: typo - "from m and SVS" should be "from SVC"

File: controller/array_action/array_mediator_svc.py [hunk 7]
```diff
@@ -294,6 +297,51 @@ def get_host_by_host_identifiers(self, initiators):
...
+            return ""
+        elif len(detailed_host_list_errors) <= MAX_HOSTS_LIST_ERR_MSG_LENGTH:
```
Comment (oriyarde) [inline controller/array_action/array_mediator_svc.py:341]: this can be shortened to:
`return textwrap.shorten(errors, width=MAX, placeholder="...")`

File: controller/array_action/svc_cli_result_reader.py [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+    Each object in response is translated to SVCListResultsElement.
+    Input is received as string which represents '\n'-separated list of returned lines from output.
+    Line with param 'id' (e.g. 'id 1') is recognized as first line of object ans used as separator between objects
+    (e.g. in input "id 1\nname n3\nid 2\nname n2" first object starts with line 'id 1' and ends with 'id 2')
```
Comment (roysahar-ibm) [inline controller/array_action/svc_cli_result_reader.py:15]: It is hard to understand the comment with the "\nn"

File: controller/array_action/array_mediator_svc.py [hunk 9]
```diff
@@ -294,6 +297,51 @@ def get_host_by_host_identifiers(self, initiators):
...
+        if not hosts_list:
+            return []
+
+        # for each host get its detailed info from array by sending batch of commands
+        detailed_hosts_list_cmd = self._get_detailed_hosts_list_cmd(hosts_list)
+        logger.debug("Sending getting detailed hosts list commands batch")
+        detailed_hosts_list_output, detailed_hosts_list_errors = self.client.send_raw_command(detailed_hosts_list_cmd)
+        detailed_hosts_list_as_string = bytes_to_string(detailed_hosts_list_output)
+        detailed_hosts_list_error_msg = self._get_formatted_hosts_list_error_msg(detailed_hosts_list_errors)
+        if not detailed_hosts_list_errors:
```
Comment (oriyarde) [inline controller/array_action/array_mediator_svc.py:312]: why `not`? perhaps we need a UT for this...

File: controller/array_action/svc_cli_result_reader.py [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+            if param_name == ID_PARAM_NAME:
+                self._next_object_id = param_value
+                return res
+            res.add(param_name, param_value)
+        return res
+
+    def _parse_param(self, line):
+        """
+        Args:
+            line : lin e representing <param-name> <param-value> (para value may be empty or contains whitespaces)
```
Comment (roysahar-ibm) [inline controller/array_action/svc_cli_result_reader.py:75]: typo - "lin e" should be "line"
"(para " should be "(param"

File: controller/array_action/svc_cli_result_reader.py (hunk 8 above)
Comment (igorgonibm) [reply controller/array_action/svc_cli_result_reader.py:15]: because it is \nname or all the comment?
If i'll use other example it ill be ok? E.g. \nwwns

File: controller/array_action/svc_cli_result_reader.py (hunk 6 above)
Comment (igorgonibm) [reply controller/array_action/svc_cli_result_reader.py:11]: fixed

File: controller/array_action/svc_cli_result_reader.py (hunk 10 above)
Comment (igorgonibm) [reply controller/array_action/svc_cli_result_reader.py:75]: fixed

File: controller/array_action/utils.py [hunk 11]
```diff
@@ -1,3 +1,10 @@
+from controller.array_action.config import UTF_8
+
+
+def bytes_to_string(input_as_bytes):
+    return input_as_bytes.decode(UTF_8) if input_as_bytes else ""
```
[the remark this answers is not in the document]
Comment (igorgonibm) [reply controller/array_action/utils.py:5]: I added to handle None value which probably may not be. But when I use this method I assume it never raises exception
This is called "defensive programming"

File: controller/tests/array_action/svc/array_mediator_svc_test.py [hunk 12]
```diff
@@ -1,15 +1,18 @@
...
+    HOST_ID_PARAM, HOST_NAME_PARAM, HOST_ISCSI_NAMES_PARAM, HOST_WWPNS_PARAM
+from controller.array_action.svc_cli_result_reader import SVCListResultsElement
+from controller.array_action.config import UTF_8
 import controller.array_action.errors as array_errors
 from pysvc.unified.response import CLIFailureError
 from pysvc import errors as svc_errors
 import controller.array_action.config as config
 from controller.common.node_info import Initiators
 
+EMPTY_BYTES = bytes("", UTF_8)
```
[the remark this answers is not in the document]
Comment (igorgonibm) [reply controller/tests/array_action/svc/array_mediator_svc_test.py:14]: done

File: controller/array_action/array_mediator_svc.py [hunk 13]
```diff
@@ -294,6 +297,51 @@ def get_host_by_host_identifiers(self, initiators):
             logger.debug("can not found host by using initiators: {0} ".format(initiators))
             raise controller_errors.HostNotFoundError(initiators)
 
+    def _get_detailed_hosts_list(self):
```
[the remark this answers is not in the document]
Comment (oriyarde) [reply controller/array_action/array_mediator_svc.py:301]: 1. no need for this general solution, since the py-cli libs don't (and has no plan to) support loading into memory in chunks.
2. the function is still too long (regardless of the general solution), and should still be divided into two.
3. as I wrote, Munch *is* different (avoids consts, avoids the wrongly-public Host class, conforms to the project's current convention, etc...).

File: controller/array_action/array_mediator_svc.py (hunk 3 above)
Comment (oriyarde) [reply controller/array_action/array_mediator_svc.py:37]: if my suggested name is better

File: controller/array_action/array_mediator_svc.py (hunk 9 above)
Comment (oriyarde) [reply controller/array_action/array_mediator_svc.py:313]: then let me rephrase: why does `if not errors` mean that "there are errors"?

File: controller/array_action/array_mediator_svc.py (hunk 1 above)
Comment (oriyarde) [reply controller/array_action/array_mediator_svc.py:332]: it is still linear to the amount of hosts (and I assume "thousands" is only a theoretical case).
so how do you know that 1 short string construction requires more instructions than 4 "append"s?
did you compare the performance between the two?

File: controller/array_action/utils.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (oriyarde) [reply controller/array_action/utils.py:6]: I understand that this could protect against future `None`.
but as I wrote, since it is only used in SVC (and only in one place), I think it's too soon to put it in common utils

File: controller/array_action/array_mediator_svc.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (igorgonibm) [reply controller/array_action/array_mediator_svc.py:301]: think the Reader is correct. It is not open for discussion.
Host object is generic. It doesn't have anything svc related and there is a ticket to generalize get_hosts_for_init where it will be used.

File: controller/array_action/utils.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (igorgonibm) [reply controller/array_action/utils.py:6]: I do not agree. Leaving this way

File: controller/array_action/array_mediator_svc.py (hunk 1 above)
Comment (igorgonibm) [reply controller/array_action/array_mediator_svc.py:332]: It doubles the time. It is not theoretical. We should support 4000 hosts.. 
Format passes over all characters and then  write passes again.

File: controller/array_action/array_mediator_svc.py (hunk 3 above)
Comment (igorgonibm) [reply controller/array_action/array_mediator_svc.py:37]: ok

File: controller/array_action/array_mediator_svc.py (hunk 4 above)
Comment (igorgonibm) [reply controller/array_action/array_mediator_svc.py:310]: ok

File: controller/array_action/array_mediator_svc.py (hunk 9 above)
Comment (igorgonibm) [reply controller/array_action/array_mediator_svc.py:313]: We don't have UT to check logs

File: controller/array_action/array_mediator_svc.py (hunk 7 above)
Comment (igorgonibm) [reply controller/array_action/array_mediator_svc.py]: seem s it doesn't work
>>textwrap.shorten('abcdefgh', width=5, placeholder="...")
'...'

Comment (oriyarde) [reply controller/array_action/array_mediator_svc.py]: my bad.
try `'{}...'.format(textwrap.wrap('abcdefgh', width=5)[0], '...')`

File: controller/array_action/array_mediator_svc.py (hunk 9 above)
Comment (oriyarde) [reply controller/array_action/array_mediator_svc.py:310]: it shouldn't be hard, but could you address my question? it's still not clear to me

File: controller/array_action/array_mediator_svc.py (hunk 1 above)
Comment (oriyarde) [reply controller/array_action/array_mediator_svc.py:329]: why double? it should also do 3 times less appends.
anyway, you can also try something like `map(writer.write, ('lshost ', host_id, ';'))`

File: controller/array_action/array_mediator_svc.py (hunk 7 above)
Comment (igorgonibm) [reply controller/array_action/array_mediator_svc.py]: This is the same thing i did except it is not one line
i pefere m y way

File: controller/array_action/array_mediator_svc.py (hunk 9 above)
Comment (igorgonibm) [reply controller/array_action/array_mediator_svc.py:310]: If not <empty list>

File: controller/array_action/utils.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (oriyarde) [reply controller/array_action/utils.py:6]: do you know if any other storage type does this conversion?
or you just think we should put storage-specific logic here?

File: controller/array_action/array_mediator_svc.py (hunk 1 above)
Comment (igorgonibm) [reply controller/array_action/array_mediator_svc.py:329]: It first "formats" pass for each char in "lshost 1" and then "writes to writer" for each char.in "lshost 1"

File: controller/array_action/config.py [hunk 14]
```diff
@@ -8,3 +8,5 @@
 
 # volume context
 CONTEXT_POOL = "pool"
+
+UTF_8 = "utf-8"
```
[the remark this answers is not in the document]
Comment (oriyarde) [reply controller/array_action/config.py:12]: why did you resolve this? I still thinks this is the wrong location for it

File: controller/array_action/array_mediator_svc.py (hunk 9 above)
Comment (oriyarde) [reply controller/array_action/array_mediator_svc.py:310]: ok. I guess it's good enough

File: controller/array_action/utils.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (igorgonibm) [reply controller/array_action/utils.py:6]: This func is not storage specific. It is general purpose func
I am stopping this discussion

File: controller/array_action/array_mediator_svc.py (hunk 1 above)
Comment (oriyarde) [reply controller/array_action/array_mediator_svc.py:329]:
```
import timeit
from io import StringIO

def _first(host_list):
	writer = StringIO()
	for host_id in host_list:
		writer.write('lshost')
		writer.write(' ')
		writer.write(host_id)
		writer.write(';')
	return writer.getvalue()


def _second(host_list):
    writer = StringIO()
    for host_id in host_list:
        writer.write('lshost {HOST_ID};'.format(HOST_ID=host_id))
    return writer.getvalue()


four_k = map(str, range(4000))

t1 = timeit.Timer("_first(four_k)", "from __main__ import _first, four_k")
first_time = t1.timeit(1)
print("1st took %fs" % (first_time,))

t2 = timeit.Timer("_second(four_k)", "from __main__ import _second, four_k")
second_time = t2.timeit(1)
print("2nd took %fs" % (second_time,))
```
1st took 0.005819s
2nd took 0.000017s

Comment (oriyarde) [reply controller/array_action/array_mediator_svc.py:329]: I'm not sure I follow the theory.
did you try to compare their time?

File: controller/array_action/array_mediator_svc.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (oriyarde) [reply controller/array_action/array_mediator_svc.py:300]: what do you mean it's not open for discussion? the review is exactly the place for these discussions.
I don't understand; couldn't find any "get_hosts_for_init" - could you point me to the ticket?

File: controller/array_action/svc_cli_result_reader.py [hunk 15]
```diff
@@ -0,0 +1,117 @@
...
+    """
+    def __init__(self, hosts_raw_list_as_string):
+        self._hosts_raw_list = hosts_raw_list_as_string.splitlines()
+        self._current_index = 0
+        self._next_object_id = None
+        self._init_first_object_id()
+
+    def _init_first_object_id(self):
+        """
+        Set _init_first_object_id to id o the first object and _current_index to point to next line
```
Comment (oriyarde) [inline controller/array_action/svc_cli_result_reader.py:25]: did you mean `id 0`? if so, does it have to be 0?

File: controller/array_action/svc_cli_result_reader.py [hunk 16]
```diff
@@ -0,0 +1,117 @@
...
+            if not line:
+                continue
+            param_name, param_value = self._parse_param(line)
+            if param_name == ID_PARAM_NAME:
+                self._next_object_id = param_value
+                return res
+            res.add(param_name, param_value)
+        return res
+
+    def _parse_param(self, line):
```
Comment (oriyarde) [inline controller/array_action/svc_cli_result_reader.py:72]: all this function for `name, _, value = line.partition(' ')`? what am I missing here?
you can also add `value = value.strip()`, but I am not aware of such need

File: controller/array_action/config.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (igorgonibm) [reply controller/array_action/config.py:12]: This is general constant. Not svc related. Even if it currently is used in svc.

File: controller/array_action/svc_cli_result_reader.py (hunk 15 above)
Comment (igorgonibm) [reply controller/array_action/svc_cli_result_reader.py:25]: no
there may be host without id 0

File: controller/array_action/svc_cli_result_reader.py (hunk 16 above)
Comment (igorgonibm) [reply controller/array_action/svc_cli_result_reader.py:72]: 1) I don't assume space separator but and whitespace
2) For empty value there ay be only single sting e.g. line "iscsi_name iscsi.ibm " or "iscsi_name" if no iscsi
3) value can contain spaces

File: controller/tests/array_action/svc/array_mediator_svc_test.py [hunk 17]
```diff
@@ -231,162 +234,158 @@ def test_properties(self):
...
-                                  'WWPN': ['abc3'],
-                                  'iscsi_name': 'iqn.test.3'})
-        ret1 = [host_munch_ret_1, host_munch_ret_2]
-        ret2 = Munch(as_single_element=host_munch_ret_2)
-        ret3 = Munch(as_single_element=host_munch_ret_3)
-        self.svc.client.svcinfo.lshost.side_effect = [ret1, ret2, ret3]
+    @patch("controller.array_action.svc_cli_result_reader.SVCListResultsReader.__iter__")
+    def test_get_host_by_identifiers_return_iscsi_host(self, result_reader_iter):
+        host_1 = self._get_host_as_dictonary('host_id_1', 'test_host_1', [], ['abc1'])
+        host_2 = self._get_host_as_dictonary('host_id_2', 'test_host_2', ['iqn.test.2'], ['abc3'])
```
Comment (roysahar-ibm) [inline controller/tests/array_action/svc/array_mediator_svc_test.py:275]: host2 is 'abc3'?

File: controller/tests/array_action/svc/array_mediator_svc_test.py [hunk 18]
```diff
@@ -231,162 +234,158 @@ def test_properties(self):
...
-        host_munch_ret_3 = Munch({'id': 'host_id_3', 'name': 'test_host_3',
-                                  'WWPN': ['abc3'],
-                                  'iscsi_name': 'iqn.test.3'})
-        ret1 = [host_munch_ret_1, host_munch_ret_2]
-        ret2 = Munch(as_single_element=host_munch_ret_2)
-        ret3 = Munch(as_single_element=host_munch_ret_3)
-        self.svc.client.svcinfo.lshost.side_effect = [ret1, ret2, ret3]
+    @patch("controller.array_action.svc_cli_result_reader.SVCListResultsReader.__iter__")
+    def test_get_host_by_identifiers_return_iscsi_host(self, result_reader_iter):
+        host_1 = self._get_host_as_dictonary('host_id_1', 'test_host_1', [], ['abc1'])
```
Comment (roysahar-ibm) [inline controller/tests/array_action/svc/array_mediator_svc_test.py:274]: host1 does not have iqn? this is for coverage of cases?

File: controller/array_action/svc_cli_result_reader.py (hunk 8 above)
Comment (igorgonibm) [reply controller/array_action/svc_cli_result_reader.py:15]: but this is the exact example. Lines with \n char separator
Do u gave any other idea how to write the example? I can do this way  but less like it
"id 1
name name1
id 2
name name2"

File: controller/array_action/svc_cli_result_reader.py (hunk 6 above)
Comment (oriyarde) [reply controller/array_action/svc_cli_result_reader.py]: if it's fixed, one of you should resolve this

File: controller/array_action/array_mediator_svc.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (oriyarde) [reply controller/array_action/array_mediator_svc.py:300]: 2. could you also address this function length issue?

File: controller/array_action/config.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (oriyarde) [reply controller/array_action/config.py:12]: even if you want single-use consts here, the way I see this file - "config" means values that we want the option to easily change in the future, and since this value won't change in python 3, we won't need to re-"configure" it.

regardless of where you eventually place it, make sure you don't duplicate the same const (e.g. `utils.py`)

File: controller/array_action/svc_cli_result_reader.py (hunk 15 above)
Comment (oriyarde) [reply controller/array_action/svc_cli_result_reader.py:25]: I still don't understand this line.
what does "Set _init_first_object_id" mean? `_init_first_object_id` is the function name, not a data member

File: controller/array_action/svc_cli_result_reader.py (hunk 16 above)
Comment (oriyarde) [reply controller/array_action/svc_cli_result_reader.py:72]: I don't think 1. is a real scenario, but I'm pretty sure my suggestion works with 2. + 3...

File: controller/tests/array_action/svc/array_mediator_svc_test.py (hunk 18 above)
Comment (igorgonibm) [reply controller/tests/array_action/svc/array_mediator_svc_test.py:274]: yes This host is without iscsi. only fc

File: controller/tests/array_action/svc/array_mediator_svc_test.py (hunk 17 above)
Comment (igorgonibm) [reply controller/tests/array_action/svc/array_mediator_svc_test.py:275]: abc3 is wwn.
This was original test value. I ddidn't change it

File: controller/tests/array_action/svc/array_mediator_svc_test.py [hunk 19]
```diff
@@ -231,162 +234,158 @@ def test_properties(self):
...
+        self.svc.client.svcinfo.lshost.return_value = self._get_hosts_list_result(hosts)
+        self.svc.client.send_raw_command = Mock()
+        self.svc.client.send_raw_command.return_value = EMPTY_BYTES, EMPTY_BYTES
+        result_reader_iter.return_value = self._get_detailed_hosts_list_result(hosts)
+        host, connectivity_type = self.svc.get_host_by_host_identifiers(Initiators('iqn.test.2', ['ABC3']))
         self.assertEqual('test_host_3', host)
         self.assertEqual([config.ISCSI_CONNECTIVITY_TYPE,
                           config.FC_CONNECTIVITY_TYPE], connectivity_type)
 
+    def _get_host_as_dictonary(self, id, name, iscsi_names_list, wwpns_list):
```
Comment (oriyarde) [inline controller/tests/array_action/svc/array_mediator_svc_test.py:359]: *dictionary

File: controller/array_action/array_mediator_svc.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (igorgonibm) [reply controller/array_action/array_mediator_svc.py:300]: Your loop doesn't handle list values. See the reader class to understand

File: controller/array_action/array_mediator_svc.py (hunk 7 above)
Comment (oriyarde) [reply controller/array_action/array_mediator_svc.py]: I agree that the logic is very similar, but what about the naming?
could you make the names more general? e.g.:
`_get_formatted_hosts_list_error_msg` -> `truncate`
`detailed_host_list_errors` -> `error_msgs`

File: controller/array_action/array_mediator_svc.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (oriyarde) [reply controller/array_action/array_mediator_svc.py:300]: please address all numbered issues (with numbers).
I think your last comment is for a different existing conversation (simply adding `value = value.split() if ' ' in value else value` after my `value = value.strip()`)

File: controller/array_action/array_mediator_svc.py (hunk 7 above)
Comment (oriyarde) [reply controller/array_action/array_mediator_svc.py]: I agree the logic is similar.
but could you still rename it to something more general? e.g.:
`_get_formatted_hosts_list_error_msg` -> `truncate_errors`
`detailed_host_list_errors` -> `errors_text`

btw I don't see this function being called anywhere.

File: controller/array_action/array_mediator_svc.py (hunk 9 above)
Comment (igorgonibm) [reply controller/array_action/array_mediator_svc.py]: fixed

Comment (oriyarde) [reply controller/array_action/array_mediator_svc.py]: thanks! but I still think the fix should include a Unit Test (to avoid this confusion in the future).
if you think it's too much for this ticket, then at least open a ticket and add it here

File: controller/array_action/array_mediator_svc.py (hunk 1 above)
Comment (oriyarde) [reply controller/array_action/array_mediator_svc.py:329]: if you want to save the time comparisons, this is the fastest and most pythonic approach:
`return 'lshost {}'.format(';lshost '.join(host_list))`

File: controller/array_action/svc_cli_result_reader.py (hunk 8 above)
Comment (igorgonibm) [reply controller/array_action/svc_cli_result_reader.py:15]: @roysahar-ibm how do u wan t me to change it?

File: controller/array_action/svc_cli_result_reader.py (hunk 15 above)
Comment (igorgonibm) [reply controller/array_action/svc_cli_result_reader.py]: @roysahar-ibm yes typo. Fixing

Comment (igorgonibm) [reply controller/array_action/svc_cli_result_reader.py]: Fixed the comment

File: controller/array_action/svc_cli_result_reader.py (hunk 16 above)
Comment (igorgonibm) [reply controller/array_action/svc_cli_result_reader.py]: done

File: controller/array_action/array_mediator_svc.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (oriyarde) [reply controller/array_action/array_mediator_svc.py:300]: yes, the loop should be in a new (second) function.
but there should also be another new (first) function for the first part (from `self.client.svcinfo.lshost()` to `if not detailed_hosts_list_errors`)

File: controller/array_action/config.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (oriyarde) [reply controller/array_action/config.py:12]: I don't remember when we talked about it, but I don't think I should open a ticket for a small rename.
if you think it is too much for this ticket, please open a new one and add the number here (we shouldn't treat `config` as `constants` without at least a ticket for renaming it)

File: controller/array_action/array_mediator_svc.py (hunk 1 above)
Comment (oriyarde) [reply controller/array_action/array_mediator_svc.py:329]: I don't understand from your last response why the writer is faster - is this about the `map` overhead?

File: controller/array_action/svc_cli_result_reader.py [hunk 20]
```diff
@@ -0,0 +1,107 @@
...
+            line = self._hosts_raw_list[self._current_index].strip()
+            self._current_index += 1
+            if line:
+                param_name, _, param_value = line.partition(' ')
+                param_value = param_value.strip()
+                if param_name == ID_PARAM_NAME:
+                    self._next_object_id = param_value
+                else:
+                    raise controller_errors.InvalidCliResponseError(
+                        "First element is {0}. Expected {1}".format(line, ID_PARAM_NAME))
```
Comment (oriyarde) [inline controller/array_action/svc_cli_result_reader.py:40]: shouldn't it be `First param is ...`? I thought "element" is the complete object, e.g. cli host

File: controller/tests/array_action/svc/array_mediator_svc_test.py [hunk 21]
```diff
@@ -231,162 +234,158 @@ def test_properties(self):
...
+        host, connectivity_type = self.svc.get_host_by_host_identifiers(Initiators('iqn.test.2', ['ABC3']))
         self.assertEqual('test_host_3', host)
         self.assertEqual([config.ISCSI_CONNECTIVITY_TYPE,
                           config.FC_CONNECTIVITY_TYPE], connectivity_type)
 
+    def _get_host_as_dictionary(self, id, name, iscsi_names_list, wwpns_list):
+        res = {}
+        res[HOST_ID_PARAM] = id
+        res[HOST_NAME_PARAM] = name
+        if iscsi_names_list is not None:
```
Comment (oriyarde) [inline controller/tests/array_action/svc/array_mediator_svc_test.py:363]: I don't think they are ever `None`.
did you mean to add `None` as a default parameter value?

File: controller/tests/array_action/svc/array_mediator_svc_test.py [hunk 22]
```diff
@@ -231,162 +234,158 @@ def test_properties(self):
...
+        self.svc.client.send_raw_command = Mock()
+        self.svc.client.send_raw_command.return_value = EMPTY_BYTES, EMPTY_BYTES
+        result_reader_iter.return_value = self._get_detailed_hosts_list_result(hosts)
+        host, connectivity_type = self.svc.get_host_by_host_identifiers(Initiators('iqn.test.2', ['ABC3']))
         self.assertEqual('test_host_3', host)
         self.assertEqual([config.ISCSI_CONNECTIVITY_TYPE,
                           config.FC_CONNECTIVITY_TYPE], connectivity_type)
 
+    def _get_host_as_dictionary(self, id, name, iscsi_names_list, wwpns_list):
+        res = {}
```
Comment (oriyarde) [inline controller/tests/array_action/svc/array_mediator_svc_test.py:360]: you can put the unconditional values in this line (`{key: value}`)

File: controller/tests/array_action/svc/array_mediator_svc_test.py [hunk 23]
```diff
@@ -231,162 +234,158 @@ def test_properties(self):
...
+        res[HOST_ID_PARAM] = id
+        res[HOST_NAME_PARAM] = name
+        if iscsi_names_list is not None:
+            res[HOST_ISCSI_NAMES_PARAM] = iscsi_names_list
+        if wwpns_list is not None:
+            res[HOST_WWPNS_PARAM] = wwpns_list
+        return res
+
+    def _get_hosts_list_result(self, hosts_dict):
```
Comment (oriyarde) [inline controller/tests/array_action/svc/array_mediator_svc_test.py:369]: remove this function and change `_get_host_as_dictionary` to `_get_host_as_munch`

File: controller/array_action/array_mediator_svc.py [hunk 24]
```diff
@@ -294,6 +297,53 @@ def get_host_by_host_identifiers(self, initiators):
...
+
+        logger.debug("Reading detailed hosts list commands batch response")
+        hosts_reader = SVCListResultsReader(detailed_hosts_list_output)
+        res = []
+        for host_details in hosts_reader:
+            host_id = host_details.get(HOST_ID_PARAM)
+            host_name = host_details.get(HOST_NAME_PARAM)
+            iscsi_names = host_details.get_as_list(HOST_ISCSI_NAMES_PARAM)
+            wwns = host_details.get_as_list(HOST_WWPNS_PARAM)
+            host = Host(host_id, host_name, iscsi_names, wwns)
```
Comment (oriyarde) [inline controller/array_action/array_mediator_svc.py:321]: opening the Host class vs Munch issue here, since the [conversation with numbered issues](https://github.com/IBM/ibm-block-csi-driver/pull/154/files#r421501818) is not convenient.
as I wrote there: could you please point me to the "get_hosts_for_init" ticket?

regardless of this ticket, I still think a Munch will save us the new consts, and conforms to the project's current convention for cli objects.

File: controller/array_action/svc_cli_result_reader.py (hunk 8 above)
Comment (roysahar-ibm) [reply controller/array_action/svc_cli_result_reader.py:15]: Replace the "\n" with "<New Line>"

Comment (igorgonibm) [reply controller/array_action/svc_cli_result_reader.py:15]: but the input contains nerwline

File: controller/array_action/array_mediator_svc.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (igorgonibm) [reply controller/array_action/array_mediator_svc.py:300]: Moved loop to separate function
but not agree with the second part. This way the original method will just call two other methods so it is meaningless.

File: controller/array_action/array_mediator_svc.py (hunk 1 above)
Comment (igorgonibm) [reply controller/array_action/array_mediator_svc.py:332]: How return 'lshost {}'.format(';lshost '.join(host_list))
do the job?

Comment (igorgonibm) [reply controller/array_action/array_mediator_svc.py:332]: I'll use: LIST_HOSTS_CMD_FORMAT = 'lshost {HOST_ID};'
writer.write(LIST_HOSTS_CMD_FORMAT.format(HOST_ID=host_id)

File: controller/array_action/array_mediator_svc.py (hunk 7 above)
Comment (igorgonibm) [reply controller/array_action/array_mediator_svc.py]: done

File: controller/array_action/array_mediator_svc.py (hunk 9 above)
Comment (igorgonibm) [reply controller/array_action/array_mediator_svc.py]: Resonsider about log assert here CSI-1492

File: controller/array_action/svc_cli_result_reader.py (hunk 20 above)
Comment (igorgonibm) [reply controller/array_action/svc_cli_result_reader.py:40]: fixed

File: controller/tests/array_action/svc/array_mediator_svc_test.py (hunk 22 above)
Comment (igorgonibm) [reply controller/tests/array_action/svc/array_mediator_svc_test.py:360]: done

File: controller/array_action/array_mediator_svc.py (hunk 24 above)
Comment (igorgonibm) [reply controller/array_action/array_mediator_svc.py:324]: consider here CSI-1492

File: controller/tests/array_action/svc/array_mediator_svc_test.py (hunk 23 above)
Comment (igorgonibm) [reply controller/tests/array_action/svc/array_mediator_svc_test.py:369]: consider CSI-1492

File: controller/tests/array_action/svc/array_mediator_svc_test.py (hunk 21 above)
Comment (igorgonibm) [reply controller/tests/array_action/svc/array_mediator_svc_test.py:363]: removed

File: controller/array_action/config.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (igorgonibm) [reply controller/array_action/config.py:12]: Removed from config

File: controller/array_action/svc_cli_result_reader.py (hunk 20 above)
Comment (oriyarde) [reply controller/array_action/svc_cli_result_reader.py]: I meant that the "element" should be replaced by "param", and not remove "param"

Comment (igorgonibm) [reply controller/array_action/svc_cli_result_reader.py]: updated

File: controller/array_action/svc_cli_result_reader.py (hunk 8 above)
Comment (igorgonibm) [reply controller/array_action/svc_cli_result_reader.py:15]: Will use <new line>

File: controller/array_action/array_mediator_svc.py (hunk 7 above)
Comment (oriyarde) [reply controller/array_action/array_mediator_svc.py]: @igorgonibm why don't you use `:` slicing?

File: controller/array_action/utils.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (oriyarde) [reply controller/array_action/utils.py:6]: I think we can avoid all this raw bytes conversion by using `self.client.transport.send_command` (instead of `self.client.send_raw_command`, similar to [the capacity type conversion avoidance trick in ds8k](https://github.com/IBM/ibm-block-csi-driver/pull/122#discussion_r381554222))
so I'm unresolving this for CSI-1492

[the remark this answers is not in the document]
Comment (oriyarde) [reply controller/array_action/utils.py:6]: fixed in https://github.com/IBM/ibm-block-csi-driver/pull/432
